datafusion/core/src/physical_plan/file_format/delimited_stream.rs

```diff
@@ -0,0 +1,215 @@
+                }
+                None => {
+                    self.remainder.extend_from_slice(&val);
+                    return;
+                }
+            },
+        };
+        let end_offset = record_ends.last().unwrap_or(start_offset);
+        if start_offset != end_offset {
+            self.complete.push_back(val.slice(start_offset..end_offset));
+        }
+
+        if end_offset != val.len() {
+            self.remainder.extend_from_slice(&val[end_offset..])
+        }
+    }
+
+    /// Marks the end of the stream, delimiting any remaining bytes
+    ///
+    /// Returns `true` if there is no remaining data to be read
+    fn finish(&mut self) -> Result<bool> {
+        if !self.remainder.is_empty() {
+            if self.is_quote {
+                return Err(DataFusionError::Execution(
+                    "encountered unterminated string".to_string(),
+                ));
+            }
+
+            if self.is_escape {
+                return Err(DataFusionError::Execution(
+                    "encountered trailing escape character".to_string(),
+                ));
+            }
+
+            self.complete
+                .push_back(Bytes::from(std::mem::take(&mut self.remainder)))
+        }
+        Ok(self.complete.is_empty())
+    }
+}
+
+impl Iterator for LineDelimiter {
+    type Item = Bytes;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.complete.pop_front()
+    }
+}
+
+/// Given a [`Stream`] of [`Bytes`] returns a [`Stream`] where each
+/// yielded [`Bytes`] contains a whole number of new line delimited records
+pub fn newline_delimited_stream<S>(s: S) -> impl Stream<Item = Result<Bytes>>
+where
+    S: Stream<Item = Result<Bytes>> + Unpin,
+{
+    let delimiter = LineDelimiter::new();
+
+    futures::stream::unfold((s, delimiter), |(mut s, mut delimiter)| async move {
+        loop {
+            if let Some(next) = delimiter.next() {
+                return Some((Ok(next), (s, delimiter)));
+            }
+
+            match s.next().await {
+                Some(Ok(bytes)) => delimiter.push(bytes),
+                Some(Err(e)) => return Some((Err(e), (s, delimiter))),
+                None => match delimiter.finish() {
+                    Ok(true) => return None,
+                    Ok(false) => continue,
+                    Err(e) => return Some((Err(e), (s, delimiter))),
+                },
+            }
+        }
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::stream::TryStreamExt;
+
+    #[test]
+    fn test_delimiter() {
+        let mut delimiter = LineDelimiter::new();
+        delimiter.push("hello\nworld");
+        delimiter.push("\n\n");
+
+        assert_eq!(delimiter.next().unwrap(), Bytes::from("hello\n"));
+        assert_eq!(delimiter.next().unwrap(), Bytes::from("world\n"));
+        assert_eq!(delimiter.next().unwrap(), Bytes::from("\n"));
+        assert!(delimiter.next().is_none());
+
+        delimiter.push("");
```

Comment:
I would recommend making a new test here as a way to make the tests more self documenting.

Unless it is important that data can be `push`ed into a `LineDelimiter` after `finish()` is called 🤔 

```suggestion
    #[test]
    fn test_delimiter_escaped() {
        delimiter.push("");
        
```

Reply:
I think it is important to test that more data can be added to a LineDelimiter after data has been pulled from it

Reply:
👍  that would be good to document (the intent of the test).

I think the main suggestion of break up the test into smaller self contained blocks with descriptive names still holds even if this particular cut-off point would not be ideal.

The total test size will be larger, but I think each test will be easier to understand what it is testing.

maybe worth a thought